### ISSUE TITLE
src/apns_connection.erl:22: Warning: undefined callback function terminate/3 (behaviour 'gen_statem')

### ISSUE THREAD

**Issue by jainygaurav** [inaka/apns4erl#243]
* Please help me resolve this warning. Please note I have a current prod env on OTP 18. I am trying to upgrade to apns4erl v2 without major other version upgrades like OTP 23 etc.

* How serious is this warning? Can I use this module with this warning? What the side-effects of this warning.

`Erlang/OTP: 19.0.7
`
```
Erlang/OTP: 19.0.7
git clone https://github.com/inaka/apns4erl.git
cd apns4erl/
wget https://s3.amazonaws.com/rebar3/rebar3 && chmod +x rebar3
./rebar3 compile
===> Fetching coveralls v2.2.0
===> Fetching jsx v2.10.0
===> Analyzing applications...
===> Compiling jsx
===> Compiling coveralls
===> Verifying dependencies...
===> Fetching base64url v1.0.1
===> Fetching gun v1.3.3
===> Fetching jsx v3.0.0
===> Fetching cowlib v2.7.3
===> Analyzing applications...
===> Compiling cowlib
===> Compiling base64url
===> Compiling gun
===> Compiling jsx
===> Analyzing applications...
===> Compiling apns
**src/apns_connection.erl:22: Warning: undefined callback function terminate/3 (behaviour 'gen_statem')**
```

**Comment by paulo-ferraz-oliveira** [comment]
https://erlang.org/doc/man/gen_statem.html was introduced in OTP 19, so that's what the compiler is complaining about. You have to either bump your OTP version or keep `apns4erl` at a lower version.

**Comment by jainygaurav** [comment]
@paulo-ferraz-oliveira I am on OTP 19.0.7 or apns4erl v2 will not compile.

from` rebar.config
`
`{minimum_otp_vsn, "19"}.
`

**Comment by paulo-ferraz-oliveira** [comment]
Sure, sorry, I read "I have a current prod env on OTP 18" and jumped to a conclusion. Lemme check.

**Comment by jainygaurav** [comment]
For your reference:

```
ls /usr/local/lib/erlang/
bin  erts-7.2.1  erts-8.0.5  Install  lib  misc  releases  usr

```

**Comment by paulo-ferraz-oliveira** [comment]
I don't get that on 19.3 (don't have 19.0 locally).

**Comment by jainygaurav** [comment]
OK...let me try 19.3

**Comment by paulo-ferraz-oliveira** [comment]
According to [the doc.s](https://erlang.org/doc/man/gen_statem.html#Module:terminate-3), not having `terminate` should not affect you much. I'm not sure if the callback became optional only after 19.0 was out (which is possible).

**Comment by jainygaurav** [comment]
@paulo-ferraz-oliveira Thank you so much. I have another issue to be resolved.

 https://github.com/inaka/apns4erl/issues/242. 

I hope with this issue, that issue is also resolved. I will update both issues accordingly after OTP 19.x upgrade and testing.

**Comment by paulo-ferraz-oliveira** [comment]
This is where it became optional: https://github.com/erlang/otp/commit/913d0b52df1e029fb1728b44ba7da318f3dc49dd.

And this points to it having become so in OTP-19.3, as per Git history.

**Comment by jainygaurav** [comment]
@paulo-ferraz-oliveira I was able to build and run apns4erl from master using OTP-19.3.6. However, https://github.com/inaka/apns4erl/issues/242 is still happening.

It happens after couple of days of non-use of apns_connection. For eg.

Last push notification sent:

`2021-03-02 21:27:36.030 [error] <0.506.0>@my_mod_apns:sendAPNS:290 Success in request. Reason was: 200 with <<"c2ca5csdsc5b0550a74fe799asda96d9049d3502c58fd5e4b27e98ed1b97e49c2d">>
`
Then after couple of days:

> 2021-03-05 15:51:38.055 [error] <0.530.0> CRASH REPORT Process <0.530.0> with 0 neighbours crashed with reason: no function clause matching gun_http2:parse(<<>>, close) line 77
> 2021-03-05 15:51:38.056 [error] <0.517.0> Supervisor gun_sup had child gun started with {gun,start_link,undefined} at <0.530.0> exit with reason function_clause in context child_terminated
> 
>

**Comment by paulo-ferraz-oliveira** [comment]
@jainygaurav, if the original issue is solved, "Warning: undefined callback", I think we should move the discussion to #242 and not mix that here. What do you say?

**Comment by jainygaurav** [comment]
Sure...I close this issue.
